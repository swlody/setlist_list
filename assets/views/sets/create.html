{% extends "layout.html" %}
{% block title %}Main page{% endblock %}
{% block header %}
  {{ super() }}
{% endblock %}
{% block content %}
  {{ super() }}
  {% block body %}
    <div class="prose p-10">
      <form hx-post="/sets" hx-ext="json-enc">
        <h1>Create new set</h1>
        <div>
          <input
            id="band-name-input"
            type="text"
            placeholder="Band name"
            name="band_name"
            class="input input-bordered w-full max-w-xs"
          />
          <input
            id="event-name-input"
            type="text"
            placeholder="Event name"
            name="event_name"
            class="input input-bordered w-full max-w-xs"
          />
          <input
            id="venue-input"
            type="text"
            placeholder="Venue"
            name="venue"
            class="input input-bordered w-full max-w-xs"
          />
          <input
            id="city-input"
            type="text"
            placeholder="City"
            name="city"
            class="input input-bordered w-full max-w-xs"
          />
          <input
            id="set-start-time-input"
            type="datetime-local"
            placeholder="Set start time"
            name="start_time"
            class="input input-bordered w-full max-w-xs"
          />
          <div id="djs-input" x-data="{ djs: 1 }">
            <template x-for="dj in djs">
              <input
                type="text"
                placeholder="DJ Name"
                name="dj_names[]"
                class="input input-bordered w-full max-w-xs"
              />
            </template>
            <button class="btn" type="button" x-show="djs < 8" x-on:click="djs++">+</button>
            <button class="btn" type="button" x-show="djs > 1" x-on:click="djs--">-</button>
          </div>
          {# TODO actually flex this #}
          <textarea
            cols="29"
            class="textarea textarea-bordered w-full max-w-xs"
            placeholder="Setlist"
          ></textarea>
        </div>
        <button class="btn" type="submit">Submit</button>
      </form>
    </div>
  {% endblock %}
{% endblock %}
