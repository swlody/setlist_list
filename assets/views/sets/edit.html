{% extends "layout.html" %}
{% block title %}Main page{% endblock %}
{% block header %}
  {{ super() }}
{% endblock %}
{% block content %}
  {{ super() }}
  {% block body %}
    <div class="prose p-10">
      <h1>Edit sets: {{ item.id }}</h1>
      <div class="mb-10">
        <form
          hx-post="/sets/{{ item.id }}"
          hx-ext="json-enc"
          hx-target="#success-message"
        >
          <div class="mb-5">
            <div>
              <label>band_name</label>
              <br />
              <input
                id="band_name"
                name="band_name"
                type="text"
                value="{{ item.band_name }}"
              />
            </div>
            <div>
              <label>date</label>
              <input
                id="date"
                name="date"
                type="date"
                value="{{ item.date }}"
              />
              <br />
            </div>
            <div>
              <label>venue</label>
              <br />
              <input
                id="venue"
                name="venue"
                type="text"
                value="{{ item.venue }}"
              />
            </div>
            <div>
              <label>setlist</label>
              <br />
              <textarea id="setlist" name="setlist" type="text">
{{ item.setlist }}</textarea
              >
            </div>
            <div>
              <div class="mt-5">
                <button
                  class="rounded-lg bg-gray-900 px-6 py-3 text-xs text-white"
                  type="submit"
                >
                  Submit
                </button>
              </div>
            </div>
          </div>
        </form>
        <!-- Open the modal using ID.showModal() method -->
        <button
          class="btn bg-red-600 px-6 py-3 text-xs text-white"
          onclick="delete_modal.showModal()"
        >
          Delete
        </button>
        <dialog id="delete_modal" class="modal">
          <div class="modal-box">
            <h3 class="text-lg font-bold">Delete</h3>
            <p class="py-4">Press ESC key or click the button below to close</p>
            <div class="modal-action">
              <form method="dialog">
                <!-- if there is a button in form, it will close the modal -->
                <button class="btn">Close</button>
                <button
                  class="btn bg-red-600 px-6 py-3 text-xs text-white"
                  hx-delete="/sets/{{ item.id }}"
                >
                  Confirm delete
                </button>
              </form>
            </div>
          </div>
        </dialog>
        <div id="success-message" class="mt-4"></div>
        <br />
        <a href="/sets">Back to sets</a>
      </div>
    </div>
  {% endblock %}
{% endblock %}
