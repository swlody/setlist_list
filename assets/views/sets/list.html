{% extends "layout.html" %}
{% block title %}Main page{% endblock %}
{% block header %}
  {{ super() }}
{% endblock %}
{% block content %}
  {{ super() }}
  {% block body %}
    <div class="prose p-10">
      <h1>Sets</h1>
      {% for item in items %}
        <div>
          <label
            ><b>DJ:</b>
            {% for dj_name in item.dj_names %}
              {{ dj_name }}
              {% if not loop.last %}B2B{% endif %}
            {% endfor %}
          </label>
          <br />
          <label><b>Start time:</b> {{ item.start_time }}</label>
          <br />
          <label><b>Venue:</b> {{ item.venue }}</label>
          <br />
          <label><b>Setlist:</b> {{ item.setlist }}</label>
          <br />
          <a href="/sets/{{ item.id }}/edit">Edit</a>
          <a href="/sets/{{ item.id }}">View</a>
        </div>
      {% endfor %}

      <br />
      <br />
      <a href="/sets/new">New set</a>
    </div>
  {% endblock %}
{% endblock %}
