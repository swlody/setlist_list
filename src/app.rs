use std::path::Path;

use async_trait::async_trait;
use axum::Router;
use loco_rs::prelude::*;
use loco_rs::{
    app::{AppContext, Hooks, Initializer},
    boot::{create_app, BootResult, StartMode},
    controller::AppRoutes,
    db::{self, truncate_table},
    environment::Environment,
    task::Tasks,
    worker::{AppWorker, Processor},
    Result,
};
use migration::Migrator;
use sea_orm::DatabaseConnection;

use crate::{
    controllers, initializers, models::_entities::users, tasks, workers::downloader::DownloadWorker,
};

pub struct App;
#[async_trait]
impl Hooks for App {
    fn app_name() -> &'static str {
        env!("CARGO_CRATE_NAME")
    }

    fn app_version() -> String {
        format!(
            "{} ({})",
            env!("CARGO_PKG_VERSION"),
            option_env!("BUILD_SHA")
                .or(option_env!("GITHUB_SHA"))
                .unwrap_or("dev")
        )
    }

    async fn boot(mode: StartMode, environment: &Environment) -> Result<BootResult> {
        create_app::<Self, Migrator>(mode, environment).await
    }

    async fn initializers(_ctx: &AppContext) -> Result<Vec<Box<dyn Initializer>>> {
        Ok(vec![Box::new(
            initializers::view_engine::ViewEngineInitializer,
        )])
    }

    fn routes(_ctx: &AppContext) -> AppRoutes {
        AppRoutes::with_default_routes()
            .add_route(controllers::index::routes())
            .add_route(controllers::auth::routes())
            .add_route(controllers::user::routes())
    }

    async fn after_routes(router: Router, _ctx: &AppContext) -> Result<Router> {
        async fn fallback_handler(
            ViewEngine(v): ViewEngine<TeraView>,
        ) -> Result<impl IntoResponse> {
            crate::views::not_found::not_found(&v)
        }

        // ref: https://github.com/loco-rs/loco/blob/1f2401951f445ef1d71ce41f562ab3d0fb89bcd3/src/controller/app_routes.rs#L375-L399
        let fallback_router = Router::new().fallback(fallback_handler).layer(
            tower_http::trace::TraceLayer::new_for_http().make_span_with(
                |request: &axum::http::Request<_>| {
                    let request_id = uuid::Uuid::new_v4();
                    let user_agent = request
                        .headers()
                        .get(axum::http::header::USER_AGENT)
                        .map_or("", |h| h.to_str().unwrap_or(""));

                    let env: String = request
                        .extensions()
                        .get::<Environment>()
                        .map(std::string::ToString::to_string)
                        .unwrap_or_default();

                    tracing::error_span!(
                        "http-request",
                        "http.method" = tracing::field::display(request.method()),
                        "http.uri" = tracing::field::display(request.uri()),
                        "http.version" = tracing::field::debug(request.version()),
                        "http.user_agent" = tracing::field::display(user_agent),
                        "environment" = tracing::field::display(env),
                        request_id = tracing::field::display(request_id),
                    )
                },
            ),
        );

        Ok(Router::new().merge(router).merge(fallback_router))
    }

    fn connect_workers<'a>(p: &'a mut Processor, ctx: &'a AppContext) {
        p.register(DownloadWorker::build(ctx));
    }

    fn register_tasks(tasks: &mut Tasks) {
        tasks.register(tasks::seed::SeedData);
    }

    async fn truncate(db: &DatabaseConnection) -> Result<()> {
        truncate_table(db, users::Entity).await?;
        Ok(())
    }

    async fn seed(db: &DatabaseConnection, base: &Path) -> Result<()> {
        db::seed::<users::ActiveModel>(db, &base.join("users.yaml").display().to_string()).await?;
        Ok(())
    }
}
