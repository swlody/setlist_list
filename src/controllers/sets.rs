use auth::JWTWithUser;
use axum::{debug_handler, http::uri::PathAndQuery};
use chrono::{NaiveDateTime, Utc};
use loco_rs::prelude::*;
use serde::{de, Deserialize, Deserializer, Serialize};
use uuid::Uuid;

use crate::{
    initializers::minijinja_view_engine::MiniJinjaView,
    models::{sets, users},
    utils::{get_username, hx_redirect},
    views,
};

fn html_datetime<'de, D>(d: D) -> Result<NaiveDateTime, D::Error>
where
    D: Deserializer<'de>,
{
    let s = String::deserialize(d)?;
    let formatted = if s.len() == 16 {
        // "yyyy-MM-ddTHH:mm" has length 16
        format!("{s}:00")
    } else {
        s
    };

    NaiveDateTime::parse_from_str(&formatted, "%Y-%m-%dT%H:%M:%S").map_err(de::Error::custom)
}

#[derive(Serialize, Deserialize)]
pub struct Params {
    pub dj_names: Vec<String>,
    pub venue: Option<String>,
    pub city: Option<String>,
    pub event_name: Option<String>,
    #[serde(deserialize_with = "html_datetime")]
    pub start_time: NaiveDateTime,
    pub duration_seconds: Option<i32>,
    pub setlist: Vec<Song>,
}

impl Params {
    fn update(self, item: &mut sets::Model) {
        item.updated_at = Utc::now().naive_utc();
        item.dj_names = self.dj_names;
        item.venue = self.venue;
        item.city = self.city;
        item.event_name = self.event_name;
        item.start_time = self.start_time;
        item.duration_seconds = self.duration_seconds;
    }
}

#[derive(Serialize, Deserialize)]
pub struct Song {
    pub track_title: String,
    pub track_artist: String,
    pub track_number: Option<i32>,
    pub track_start_time_offset_seconds: Option<i32>,
    pub track_duration_seconds: Option<i32>,
}

async fn load_item(ctx: &AppContext, id: Uuid) -> Result<sets::Model> {
    Ok(sets::Model::find_by_id(&ctx.db, id).await?)
}

#[debug_handler]
pub async fn list(
    jwt_user: Option<JWTWithUser<users::Model>>,
    ViewEngine(v): ViewEngine<MiniJinjaView>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user_name = get_username(jwt_user).unwrap_or_default();
    let item = sets::Model::list_all(&ctx.db).await;
    if let Ok(item) = item {
        views::sets::list(&v, &item, &user_name)
    } else {
        views::index::not_found(&v, &user_name)
    }
}

#[debug_handler]
pub async fn new(
    jwt_user: Option<JWTWithUser<users::Model>>,
    ViewEngine(v): ViewEngine<MiniJinjaView>,
    State(_ctx): State<AppContext>,
) -> Result<Response> {
    let user_name = get_username(jwt_user).unwrap_or_default();
    if user_name.is_empty() {
        views::index::unauthorized(&v)
    } else {
        views::sets::create(&v, &user_name)
    }
}

#[debug_handler]
pub async fn update(
    Path(id): Path<Uuid>,
    State(ctx): State<AppContext>,
    Json(params): Json<Params>,
) -> Result<Response> {
    let mut item = load_item(&ctx, id).await?;
    params.update(&mut item);
    item.update(&ctx.db).await?;
    format::json(item)
}

#[debug_handler]
pub async fn edit(
    jwt_user: Option<JWTWithUser<users::Model>>,
    Path(id): Path<Uuid>,
    ViewEngine(v): ViewEngine<MiniJinjaView>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user_name = get_username(jwt_user).unwrap_or_default();
    if user_name.is_empty() {
        views::index::unauthorized(&v)
    } else {
        let item = load_item(&ctx, id).await;
        if let Ok(item) = item {
            views::sets::edit(&v, &item, &user_name)
        } else {
            views::index::not_found(&v, &user_name)
        }
    }
}

#[debug_handler]
pub async fn show(
    jwt_user: Option<JWTWithUser<users::Model>>,
    path: Option<Path<Uuid>>,
    ViewEngine(v): ViewEngine<MiniJinjaView>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user_name = get_username(jwt_user).unwrap_or_default();
    if let Some(Path(id)) = path {
        let item = load_item(&ctx, id).await;
        if let Ok(item) = item {
            views::sets::show(&v, &item, &user_name)
        } else {
            views::index::not_found(&v, &user_name)
        }
    } else {
        views::index::not_found(&v, &user_name)
    }
}

#[debug_handler]
pub async fn add(
    auth: auth::JWT,
    State(ctx): State<AppContext>,
    Json(params): Json<Params>,
) -> Result<Response> {
    let uuid = uuid::Uuid::parse_str(&auth.claims.id)?;
    let mut item = sets::Model {
        id: Uuid::now_v7(),
        creator_id: uuid,
        ..Default::default()
    };
    params.update(&mut item);
    item.insert(&ctx.db).await?;
    hx_redirect(&PathAndQuery::from_static("/sets"))
}

#[debug_handler]
pub async fn remove(Path(id): Path<Uuid>, State(ctx): State<AppContext>) -> Result<Response> {
    sets::Model::delete_by_id(&ctx.db, id).await?;
    hx_redirect(&PathAndQuery::from_static("/sets"))
}

pub fn routes() -> Routes {
    Routes::new()
        .prefix("sets")
        .add("/", get(list))
        .add("/new", get(new))
        .add("/:id", get(show))
        .add("/:id/edit", get(edit))
        .add("/:id", post(update))
        .add("/:id", delete(remove))
        .add("/", post(add))
}
