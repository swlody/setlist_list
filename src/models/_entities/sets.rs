//! `SeaORM` Entity. Generated by sea-orm-codegen 0.12.15

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "sets")]
pub struct Model {
    pub created_at: DateTime,
    pub updated_at: DateTime,
    #[sea_orm(primary_key)]
    pub id: i32,
    pub band_name: Option<String>,
    pub date: ChronoDate,
    pub venue: Option<String>,
    pub setlist: Option<Json>,
    pub creator_pid: Uuid,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}
