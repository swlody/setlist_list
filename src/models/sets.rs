use chrono::NaiveDateTime;
use loco_rs::model::{ModelError, ModelResult};
use serde::{Deserialize, Serialize};
use sqlx::{types::Uuid, PgPool};

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Default)]
pub struct Model {
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
    pub id: Uuid,
    pub creator_id: Uuid,
    pub dj_names: Vec<String>,
    pub venue: Option<String>,
    pub city: Option<String>,
    pub event_name: Option<String>,
    pub start_time: NaiveDateTime,
    pub duration_seconds: Option<i32>,
}

impl Model {
    pub async fn list_by_creator_id(db: &PgPool, id: Uuid) -> ModelResult<Vec<Self>> {
        Ok(
            sqlx::query_as!(Self, "SELECT * FROM sets WHERE creator_id = $1", id)
                .fetch_all(db)
                .await?,
        )
    }

    pub async fn find_by_id(db: &PgPool, id: Uuid) -> ModelResult<Self> {
        let set = sqlx::query_as!(Self, "SELECT * FROM sets WHERE id = $1", id)
            .fetch_optional(db)
            .await?;
        set.ok_or(ModelError::EntityNotFound)
    }

    pub async fn delete_by_id(db: &PgPool, id: Uuid) -> ModelResult<()> {
        sqlx::query!("DELETE FROM sets WHERE id = $1", id)
            .execute(db)
            .await?;
        Ok(())
    }

    pub async fn list_all(db: &PgPool) -> ModelResult<Vec<Self>> {
        Ok(sqlx::query_as!(Self, "SELECT * FROM sets")
            .fetch_all(db)
            .await?)
    }

    pub async fn insert(&self, db: &PgPool) -> ModelResult<()> {
        sqlx::query!(
            r#"INSERT INTO sets (id, creator_id, dj_names, venue, city, event_name, start_time, duration_seconds)
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8)"#,
            self.id,
            self.creator_id,
            &self.dj_names,
            self.venue,
            self.city,
            self.event_name,
            self.start_time,
            self.duration_seconds
        )
        .execute(db)
        .await?;
        Ok(())
    }

    pub async fn update(&self, db: &PgPool) -> ModelResult<()> {
        sqlx::query!(
            r#"UPDATE sets
            SET dj_names = $1, venue = $2, city = $3, event_name = $4, start_time = $5, duration_seconds = $6"#,
            &self.dj_names,
            self.venue,
            self.city,
            self.event_name,
            self.start_time,
            self.duration_seconds,
        )
        .execute(db)
        .await?;
        Ok(())
    }
}
