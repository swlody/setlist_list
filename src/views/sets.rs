use loco_rs::prelude::*;

use crate::models::_entities::sets;

/// Render a list view of sets.
///
/// # Errors
///
/// When there is an issue with rendering the view.
pub fn list(v: &impl ViewRenderer, items: &Vec<sets::Model>, user_name: &str) -> Result<Response> {
    format::render().view(
        v,
        "sets/list.html",
        serde_json::json!({"items": items, "username": user_name}),
    )
}

/// Render a single sets view.
///
/// # Errors
///
/// When there is an issue with rendering the view.
pub fn show(v: &impl ViewRenderer, item: &sets::Model, user_name: &str) -> Result<Response> {
    format::render().view(
        v,
        "sets/show.html",
        serde_json::json!({"item": item, "username": user_name}),
    )
}

/// Render a sets create form.
///
/// # Errors
///
/// When there is an issue with rendering the view.
pub fn create(v: &impl ViewRenderer, user_name: &str) -> Result<Response> {
    format::render().view(
        v,
        "sets/create.html",
        serde_json::json!({"username": user_name}),
    )
}

/// Render a sets edit form.
///
/// # Errors
///
/// When there is an issue with rendering the view.
pub fn edit(v: &impl ViewRenderer, item: &sets::Model, user_name: &str) -> Result<Response> {
    format::render().view(
        v,
        "sets/edit.html",
        serde_json::json!({"item": item, "username": user_name}),
    )
}
