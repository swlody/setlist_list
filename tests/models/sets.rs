use insta::assert_debug_snapshot;
use loco_rs::testing;
use setlist_list::app::App;
use setlist_list::models::sets::Model;
use sqlx::PgPool;
use uuid::uuid;

macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        let _guard = settings.bind_to_scope();
    };
}

#[sqlx::test(fixtures("sets"))]
async fn can_find_by_id(pool: PgPool) -> eyre::Result<()> {
    configure_insta!();

    let boot = testing::boot_test::<App>(pool).await?;

    let item = Model::find_by_id(
        &boot.app_context.db,
        uuid!("33333333-3333-3333-3333-333333333333"),
    )
    .await?;

    // snapshot the result:
    assert_debug_snapshot!(item);
    Ok(())
}
