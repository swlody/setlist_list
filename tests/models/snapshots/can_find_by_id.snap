---
source: tests/models/sets.rs
expression: item
---
Model {
    created_at: 2022-06-02T11:00:00,
    updated_at: 2022-06-02T11:20:00,
    id: 33333333-3333-3333-3333-333333333333,
    creator_id: 11111111-1111-1111-1111-111111111111,
    dj_names: [
        "Jeff Mills",
        "Frankie Knuckles",
    ],
    venue: Some(
        "Marble Bar",
    ),
    city: Some(
        "Detroit",
    ),
    event_name: Some(
        "Movement Festival",
    ),
    start_time: 2022-05-27T11:00:00,
    setlist: Some(
        Object {},
    ),
}
